In [2]:
pip install transformers torch sentencepiece

In [3]:
# Unit 1 – Model Benchmarking Assignment
### Models: BERT vs RoBERTa vs BART
### Objective: Observe model behavior beyond intended architecture

In [4]:
from transformers import pipeline

In [5]:
#BERT Text Generation
try:
    bert_gen = pipeline("text-generation", model="bert-base-uncased")
    output = bert_gen("The future of Artificial Intelligence is", max_length=30)
    print(output)
except Exception as e:
    print("BERT Error:", e)

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

If you want to use `BertLMHeadModel` as a standalone, add `is_decoder=True.`


tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

Device set to use cpu
Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Both `max_new_tokens` (=256) and `max_length`(=30) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


[{'generated_text': 'The future of Artificial Intelligence is................................................................................................................................................................................................................................................................'}]


In [6]:
#RoBERTa Text Generation
try:
    roberta_gen = pipeline("text-generation", model="roberta-base")
    output = roberta_gen("The future of Artificial Intelligence is", max_length=30)
    print(output)
except Exception as e:
    print("RoBERTa Error:", e)

config.json:   0%|          | 0.00/481 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/499M [00:00<?, ?B/s]

If you want to use `RobertaLMHeadModel` as a standalone, add `is_decoder=True.`


tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

Device set to use cpu
Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Both `max_new_tokens` (=256) and `max_length`(=30) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


[{'generated_text': 'The future of Artificial Intelligence is'}]


In [7]:
#BART Text Generation
bart_gen = pipeline("text-generation", model="facebook/bart-base")
output = bart_gen("The future of Artificial Intelligence is", max_length=30)
print(output)

config.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/558M [00:00<?, ?B/s]

Some weights of BartForCausalLM were not initialized from the model checkpoint at facebook/bart-base and are newly initialized: ['lm_head.weight', 'model.decoder.embed_tokens.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cpu
Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Both `max_new_tokens` (=256) and `max_length`(=30) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


[{'generated_text': 'The future of Artificial Intelligence is 157 157 157 Dickinson 157 157iled 157 157ma 157 157 earthquakeshammad 157ocket 157oola 157Mark 157 157 TXUUiledhammadMark earthquakes 157 157 retake 157 157 Karen 157anch earthquakes 157 attractionsU 157 157Unn rent 157 157 Ivan 157iled earthquakes 157 computersiled 157 satisfy 157 157 believesU 157iled broken 157 157い Moe 157 157 Found 157 157Reuters 157 (£ 157 earthquakes 157 believes woods 157 Diary earthquakes manipulating 157 157 brill 157 earthquakes earthquakes 157classesclasses earthquakes 157 earthquakesclasses 157 157 graduate earthquakes earthquakes earthquakes reliabilitycomm 157 Freder 157 157 cat earthquakes Strange 157 graduateclasses earthquakes earthquakesclasses earthquakesBrand 157 157 Davies 157 Cornell 157Uclasses 157 earthquakesiled (£classes earthquakesclassesinky hyp 157 earthquakes cat overc 157 Found earthquakes 157 graduate Found Foundclassesvaccclassesclassesractions earthquakesclassesclasses grad

In [8]:
#BERT Fill-Mask
bert_mask = pipeline("fill-mask", model="bert-base-uncased")
bert_mask("The goal of Generative AI is to [MASK] new content.")

Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForMaskedLM: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Device set to use cpu


[{'score': 0.5396932363510132,
  'token': 3443,
  'token_str': 'create',
  'sequence': 'the goal of generative ai is to create new content.'},
 {'score': 0.15575720369815826,
  'token': 9699,
  'token_str': 'generate',
  'sequence': 'the goal of generative ai is to generate new content.'},
 {'score': 0.05405500903725624,
  'token': 3965,
  'token_str': 'produce',
  'sequence': 'the goal of generative ai is to produce new content.'},
 {'score': 0.04451530799269676,
  'token': 4503,
  'token_str': 'develop',
  'sequence': 'the goal of generative ai is to develop new content.'},
 {'score': 0.01757744885981083,
  'token': 5587,
  'token_str': 'add',
  'sequence': 'the goal of generative ai is to add new content.'}]

In [9]:
#RoBERTa Fill Mask
roberta_mask = pipeline("fill-mask", model="roberta-base")
roberta_mask("The goal of Generative AI is to <mask> new content.")

Device set to use cpu


[{'score': 0.3711312413215637,
  'token': 5368,
  'token_str': ' generate',
  'sequence': 'The goal of Generative AI is to generate new content.'},
 {'score': 0.3677145540714264,
  'token': 1045,
  'token_str': ' create',
  'sequence': 'The goal of Generative AI is to create new content.'},
 {'score': 0.08351420611143112,
  'token': 8286,
  'token_str': ' discover',
  'sequence': 'The goal of Generative AI is to discover new content.'},
 {'score': 0.021335121244192123,
  'token': 465,
  'token_str': ' find',
  'sequence': 'The goal of Generative AI is to find new content.'},
 {'score': 0.016521666198968887,
  'token': 694,
  'token_str': ' provide',
  'sequence': 'The goal of Generative AI is to provide new content.'}]

In [10]:
#BART Fill Mask
bart_mask = pipeline("fill-mask", model="facebook/bart-base")
bart_mask("The goal of Generative AI is to <mask> new content.")

Device set to use cpu


[{'score': 0.07461541891098022,
  'token': 1045,
  'token_str': ' create',
  'sequence': 'The goal of Generative AI is to create new content.'},
 {'score': 0.06571870297193527,
  'token': 244,
  'token_str': ' help',
  'sequence': 'The goal of Generative AI is to help new content.'},
 {'score': 0.060880109667778015,
  'token': 694,
  'token_str': ' provide',
  'sequence': 'The goal of Generative AI is to provide new content.'},
 {'score': 0.03593561053276062,
  'token': 3155,
  'token_str': ' enable',
  'sequence': 'The goal of Generative AI is to enable new content.'},
 {'score': 0.03319477662444115,
  'token': 1477,
  'token_str': ' improve',
  'sequence': 'The goal of Generative AI is to improve new content.'}]

In [11]:
#BERT Question Answering
qa_bert = pipeline("question-answering", model="bert-base-uncased")

qa_bert(
    question="What are the risks?",
    context="Generative AI poses significant risks such as hallucinations, bias, and deepfakes."
)

Some weights of BertForQuestionAnswering were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['qa_outputs.bias', 'qa_outputs.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Device set to use cpu


{'score': 0.011546223424375057, 'start': 72, 'end': 81, 'answer': 'deepfakes'}

In [12]:
#RoBERTa QA
qa_roberta = pipeline("question-answering", model="roberta-base")

qa_roberta(
    question="What are the risks?",
    context="Generative AI poses significant risks such as hallucinations, bias, and deepfakes."
)

Some weights of RobertaForQuestionAnswering were not initialized from the model checkpoint at roberta-base and are newly initialized: ['qa_outputs.bias', 'qa_outputs.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Device set to use cpu


{'score': 0.01303130341693759,
 'start': 43,
 'end': 81,
 'answer': 'as hallucinations, bias, and deepfakes'}

In [13]:
#BART QA
qa_bart = pipeline("question-answering", model="facebook/bart-base")

qa_bart(
    question="What are the risks?",
    context="Generative AI poses significant risks such as hallucinations, bias, and deepfakes."
)

Some weights of BartForQuestionAnswering were not initialized from the model checkpoint at facebook/bart-base and are newly initialized: ['qa_outputs.bias', 'qa_outputs.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Device set to use cpu


{'score': 0.02577572502195835,
 'start': 46,
 'end': 66,
 'answer': 'hallucinations, bias'}

In [ ]:
#Observation Table

| Task           | Model       | Classification (Success/Failure) | Observation (What actually happened?)                                                           | Why did this happen? (Architectural Reason)                                                                                                                       |
| -------------- | ----------- | -------------------------------- | ----------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **Generation** | **BERT**    | Failure                          | Generated repetitive punctuation (`.....`) instead of meaningful text.                          | BERT is an encoder-only model trained with Masked Language Modeling and is not designed to predict the next word in a sequence.                                   |
| **Generation** | **RoBERTa** | Failure                          | Failed to generate a coherent continuation of the sentence.                                     | RoBERTa is also an encoder-only architecture focused on understanding tasks and lacks a decoder for autoregressive text generation.                               |
| **Generation** | **BART**    | Partial / Failure                | Generated incoherent output with repeated tokens such as `157`, resulting in gibberish text.    | BART is an encoder–decoder model intended for sequence-to-sequence tasks; forcing it into causal text generation leads to unstable decoding and token repetition. |
| **Fill-Mask**  | **BERT**    | Success                          | Correctly predicted words like *create*, *generate*, and *produce* with high confidence.        | BERT is trained using Masked Language Modeling (MLM), making it well-suited for predicting masked tokens.                                                         |
| **Fill-Mask**  | **RoBERTa** | Success                          | Successfully predicted meaningful verbs such as *generate* and *create* with strong confidence. | RoBERTa is optimized for MLM with improved training strategies, leading to strong fill-mask performance.                                                          |
| **Fill-Mask**  | **BART**    | Partial                          | Predicted reasonable but generic verbs with much lower confidence than BERT and RoBERTa.        | BART uses a denoising autoencoder objective rather than word-level MLM, so fill-mask is not its primary strength.                                                 |
| **QA**         | **BERT**    | Partial                          | Returned a single keyword (*deepfakes*) with very low confidence.                               | The base BERT model is not fine-tuned for question answering; QA requires task-specific fine-tuning such as SQuAD.                                                |
| **QA**         | **RoBERTa** | Partial                          | Extracted a longer phrase (*hallucinations, bias, and deepfakes*) but with low confidence.      | Although strong in language understanding, RoBERTa needs QA-specific fine-tuning to perform reliably on question answering tasks.                                 |
| **QA**         | **BART**    | Partial                          | Returned an incomplete answer (*hallucinations, bias*) and missed part of the context.          | The base BART model is not fine-tuned for QA, resulting in weak answer extraction.                                                                                |
